### PR TITLE
Add forms core

### PR DESCRIPTION
**Pull request by rafenden**

### CODE REVIEW

File: src/forms/core/Form.jsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  const { children, ...rest } = props
+  const formContextProps = useFormContext(rest)
+  const { submitForm } = formContextProps
+
+  const renderChildren = () => {
+    if (typeof children === 'function') {
+      return children(formContextProps)
+    }
+    return children
+  }
```
Comment (paulgain) [inline src/forms/core/Form.jsx]: Consider a one-liner:
`return (typeof children === 'function') ? : children(formContextProps) : children`

File: src/forms/journeys/MultistepExampleForm.jsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  return (
+    <>
+      <pre>{JSON.stringify(form, null, 2)}</pre>
+    </>
+  )
+}
+
+function sendToBackend(values) {
+  // eslint-disable-next-line no-alert
+  alert(JSON.stringify(values))
```
Comment (paulgain) [inline src/forms/journeys/MultistepExampleForm.jsx]: Left in on purpose?

Comment (rafenden) [reply src/forms/journeys/MultistepExampleForm.jsx]: Yes, this is just an example to help people understand how to create forms.

Will be removed once we have some forms created.

Comment (paulgain) [reply src/forms/journeys/MultistepExampleForm.jsx]: 👍

File: src/forms/elements/FieldInput.jsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      <input
+        {...rest}
+        name={name}
+        key={name}
+        id={name}
+        type={type}
+        value={value}
+        onChange={onChange}
+        onBlur={onBlur}
+        style={error ? { border: 'solid 1px red' } : null}
```
Comment (paulgain) [inline src/forms/elements/FieldInput.jsx]: Consider importing the colour for consistency: `import { RED } from 'govuk-colours'`?

File: src/forms/core/__tests__/Form.test.jsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import React from 'react'
+import { mount } from 'enzyme'
+import Form from '../Form'
+import { FieldInput } from '../../elements'
+
+describe('Form', () => {
+  describe('when form has many fields', () => {
+    test('should render without issues', () => {
```
Comment (paulgain) [inline src/forms/core/__tests__/Form.test.jsx]: Would be better to say `should render 14 fields without issue`

File: src/forms/core/useFormContext.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
+import { useState } from 'react'
+import createUseContext from 'constate'
+import { isEmpty } from 'lodash'
+
```
Comment (paulgain) [inline src/forms/core/useFormContext.js]: Looks like a good example of how to use React Hooks. 👍

**Comment by paulgain** [comment]
@rafenden looks good. I've left a few minor comments. It would be nice to see a README on how to create a Form with and without Steps despite it being clearly defined within the unit tests.

File: src/forms/core/Form.jsx (hunk 1 above)
Comment (rafenden) [reply src/forms/core/Form.jsx]: I think it's slightly easier to read in the current form.

Comment (yeahfro) [reply src/forms/core/Form.jsx]: Just to add another option using Lodash's `isFunction` 😂 

```javascript
return isFunction(children) ? children(formContextProps) : children
```

File: src/forms/core/useForm.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      const newSteps = [...prevSteps]
+      newSteps.splice(currentStep + 1, 0, name)
+      return newSteps
+    }
+    return prevSteps
+  })
+  const deregisterStep = name => setSteps(prevSteps => prevSteps.filter(s => s !== name))
+
+  const goToNextStep = () => {
+    const validationErrors = validateForm()
```
Comment (yeahfro) [inline src/forms/core/useForm.js]: If I am reading this step navigation correctly then it implies that we move through the steps in sequence. This is okay for now but we will need to understand that when we have more complex journeys then the user could branch in a direction.

File: src/forms/core/__tests__/Step.test.jsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      expect(wrapper.find('button').text()).toEqual('Next')
+      expect(formState.currentStep).toEqual(0)
+      expect(formState.steps).toEqual([
+        'testStep1',
+        'testStep2',
+      ])
+      expect(Object.keys(formState.fields)).toEqual(['testField1'])
+    })
+
+    test('should render second step when Next button is pressed', () => {
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/forms/core/__tests__/Step.test.jsx]: Are you suggesting to keep `when there are more steps` as parent?

```
describe('when there are more steps')
  describe('when the Next button is pressed')
    ...
    test('...')
```

[the remark this answers is not in the document]
Comment (yeahfro) [reply src/forms/core/__tests__/Step.test.jsx]: Yeah that's right

**Pull request merged**